### ISSUE TITLE
Alternative CoreML inference approach for macOS users (Reduces processing time by 85%)

### ISSUE THREAD

**Issue by happpycorn** [Steffenhir/GraXpert#252]
# Alternative CoreML inference approach for macOS users (Reduces processing time by 85%)

## Overview

I have been experimenting with using **CoreML** instead of **ONNX Runtime** to run the GraXpert denoise model. This modification allows the model to fully utilize the GPU/Apple Neural Engine on my MacBook.

The performance improvement is massive: it achieves a **6x to 7x speedup (600%~700% faster)** compared to the standard CPU fallback.

## Test Environment

* Mac model: MacBook Air M4
* Siril script version: 2.0.3
* GraXpert AI model: 3.0.2

## Important Notes

1. **Siril Script Only:** I applied this modification to the Siril integration script (`GraXpert-AI.py`) because it was easier for me to tweak. I'm not entirely sure how to replicate this directly in the main GraXpert GUI app.
2. **Not tested on BGE/Deconv:** The same logic should technically apply to Background Extraction and Deconvolution, but I only modified the Denoise function for this test (BGE is already fast enough for me).
3. **Use at your own risk:** I haven't fully tested this for potential bugs or unexpected consequences.
4. **Just a temporary alternative:** This is obviously a hacky workaround and not a maintainable long-term solution. I just wanted to share the idea so other Mac users can enjoy the speedup right now. If anyone has ideas on how to properly integrate CoreML support into GraXpert, I'd love to see it!

### 1. Find the ONNX model on your computer

For my setup, the default path to the downloaded model is:
`/Users/<your_username>/Library/Application Support/GraXpert/denoise-ai-models/3.0.2/model.onnx`

### 2. Convert the ONNX Model to CoreML

Run the following Python script to convert the ONNX model into a CoreML package with dynamic batching support.

*(Note: On macOS, a `.mlpackage` is actually a directory, not a single file).*

```python
import torch
import coremltools as ct
from onnx2torch import convert

# Load ONNX and convert to PyTorch
onnx_file_path = "<path_to_your_model.onnx>"
pytorch_model = convert(onnx_file_path)
pytorch_model.eval()

# Set dynamic batch dimensions (1 to 64)
batch_range = ct.RangeDim(lower_bound=1, upper_bound=64, default=1)
input_shape = ct.Shape(shape=(batch_range, 256, 256, 3))

# Trace the model with dummy data
dummy_input = torch.randn(1, 256, 256, 3)
traced_model = torch.jit.trace(pytorch_model, dummy_input)

# Convert and save
mlmodel = ct.convert(
    traced_model,
    inputs=[ct.TensorType(shape=input_shape, name="input_1")],
    convert_to="mlprogram"
)
mlmodel.save("model.mlpackage")
```

(Optional) You can use this script to test the performance of the generated CoreML model across different batch sizes:

```python
import coremltools as ct
import numpy as np
import time

model_path = "model.mlpackage"
model = ct.models.MLModel(model_path, compute_units=ct.ComputeUnit.ALL)
input_name = model.get_spec().description.input[0].name

batch_sizes = [1, 4, 8, 16, 32, 64]
num_runs = 20

print("Starting throughput test...")

for batch_size in batch_sizes:
    dummy_input = np.random.randn(batch_size, 256, 256, 3).astype(np.float32)
    
    # Warm-up
    model.predict({input_name: dummy_input})
    
    start_time = time.time()
    for _ in range(num_runs):
        model.predict({input_name: dummy_input})
    end_time = time.time()
    
    avg_batch_time_ms = ((end_time - start_time) * 1000) / num_runs
    avg_per_image_time_ms = avg_batch_time_ms / batch_size
    
    print(f"Batch Size: {batch_size:2d} | Total: {avg_batch_time_ms:.2f} ms | Per Image: {avg_per_image_time_ms:.2f} ms")
```

### 3. Copy the CoreML package to the models directory

Next, you need to place the generated `model.mlpackage` into the correct location so the script can find it.

Create a new folder inside the `denoise-ai-models` directory (whatever name you choose for this folder is what will appear in the Siril GUI drop-down menu), and paste the package inside.

Here is an example of the structure:

```text
denoise-ai-models/
├── 3.0.2/
│   └── model.onnx       # Your existing model
└── CoreML/              # You need to create this folder
    └── model.mlpackage  # The folder created in Step 2
```

### 4. Patch the Siril Script (GraXpert-AI.py)

Open your local Siril script. For macOS users, the default path is usually: `/Users/<your_username>/Library/Application Support/org.siril.Siril/siril-scripts/processing/GraXpert-AI.py`

Open this file in any text editor and make the following modifications:

#### 4a. Ensure coremltools is installed automatically

Find the `s.ensure_installed(...)` function near the top of the script (around line 90) and add "coremltools" to the list.

```python
# Before
s.ensure_installed("numpy", "astropy", "appdirs",
                   "opencv-python", "PyQt6")
```

```python
# After
s.ensure_installed("numpy", "astropy", "appdirs",
                   "opencv-python", "PyQt6", "coremltools")
```

#### 4b. Allow the script to recognize .mlpackage folders

Find the `get_available_local_models` function and replace the model-checking loop with the following:

```python
# Before
        for subdir in os.listdir(models_dir):
            subdir_path = os.path.join(models_dir, subdir)
            if os.path.isdir(subdir_path):
                model_path = os.path.join(subdir_path, "model.onnx")
                if os.path.exists(model_path) and os.path.isfile(model_path):
                    # Use subdirectory name as the display name
                    model_paths[subdir] = model_path
```

```python
# After
        for subdir in os.listdir(models_dir):
            subdir_path = os.path.join(models_dir, subdir)
            if os.path.isdir(subdir_path):
                coreml_path = os.path.join(subdir_path, "model.mlpackage")
                onnx_path = os.path.join(subdir_path, "model.onnx")
                
                if os.path.exists(coreml_path) and os.path.isdir(coreml_path):
                    model_paths[subdir] = coreml_path
                elif os.path.exists(onnx_path) and os.path.isfile(onnx_path):
                    model_paths[subdir] = onnx_path
```

#### 4c. Fix path validation (Crucial for macOS directories)

Since a `.mlpackage` is actually a directory on macOS, `os.path.isfile()` will fail and block the execution. Find `GUIInterface._on_apply` and replace the validation logic:

```python
# Before
        # Validate model path
        if not model_path or not os.path.isfile(model_path):
            QMessageBox.critical(self, "Error", "Please select a valid ONNX model file.")
            return
```

```python
# After
        # Validate model path
        if not model_path or not os.path.exists(model_path):
            QMessageBox.critical(self, "Error", "Please select a valid ONNX model file.")
```

#### 4d. Fix path validation (Same as 4c)

Find the `check_config_file` function inside `DenoiserProcessing` (and ideally in `DeconvolutionProcessing` and `BGEProcessing` as well) and apply the same path validation fix.

(Fun fact: If you forget to modify this, the script will miraculously still work :D)

```python
# Before
                    if not os.path.isfile(model_path):
                        model_path = None
```

```python
# After
                    if not os.path.exists(model_path):
                        model_path = None
```

#### 4e. Update the Inference Engine Initialization

Find the `# Initialize ONNX runtime session` block inside `DenoiserProcessing.denoise` and modify it to support both `.mlpackage` and `.onnx` formats.

```python
# Before
        # Initialize ONNX runtime session
        with s.SuppressedStderr():
            providers = onnx_helper.get_execution_providers_ordered(ai_gpu_acceleration)
            try:
                session = onnxruntime.InferenceSession(ai_path, providers=providers)
            except Exception as err:
                error_message = str(err)
                print("Warning: falling back to CPU.")
                if "cudaErrorNoKernelImageForDevice" in error_message \
                    or "Error compiling model" in error_message:
                    print("ONNX cannot build an inferencing kernel for this GPU.")
                # Retry with CPU only
                providers = ['CPUExecutionProvider']
                try:
                    session = onnxruntime.InferenceSession(ai_path, providers=providers)
                except ONNXRuntimeError as err:
                    messagebox.showerror("Error", "Cannot build an inference model on this device")
                    return

        print(f"Using inference providers: {session.get_providers()}")
```

```python
# After
        is_coreml = ai_path.endswith('.mlpackage')
        
        if is_coreml:
            import coremltools as ct
            session = ct.models.MLModel(ai_path, compute_units=ct.ComputeUnit.ALL)
            coreml_input_name = session.get_spec().description.input[0].name
            coreml_output_name = session.get_spec().description.output[0].name
        else:
            with s.SuppressedStderr():
                providers = onnx_helper.get_execution_providers_ordered(ai_gpu_acceleration)
                try:
                    session = onnxruntime.InferenceSession(ai_path, providers=providers)
                except Exception as err:
                    error_message = str(err)
                    print("Warning: falling back to CPU.")
                    if "cudaErrorNoKernelImageForDevice" in error_message \
                        or "Error compiling model" in error_message:
                        print("ONNX cannot build an inferencing kernel for this GPU.")
                    # Retry with CPU only
                    providers = ['CPUExecutionProvider']
                    try:
                        session = onnxruntime.InferenceSession(ai_path, providers=providers)
                    except ONNXRuntimeError as err:
                        messagebox.showerror("Error", "Cannot build an inference model on this device")
                        return

            print(f"Using inference providers: {session.get_providers()}")
```

#### 4f. Update the Inference Execution

Find the `# Run inference section` inside `DenoiserProcessing.denoise`. We need to implement a branch so the script uses `session.predict` for CoreML models and the original `onnx_helper.run` for ONNX models.

```python
# Before
            # Run inference
            output_tiles = []

            session_result, session = onnx_helper.run(session, ai_path, None, \
                        {"gen_input_image": input_tiles}, return_first_output=True)

            for e in session_result:
                output_tiles.append(e)
```

```python
# After
            # Run inference
            output_tiles = []

            if is_coreml:
                # CoreML requires float32 NumPy arrays
                input_tiles_fp32 = input_tiles.astype(np.float32)
                predictions = session.predict({coreml_input_name: input_tiles_fp32})
                session_result = predictions[coreml_output_name]
            else:
                # Original ONNX inference
                session_result, session = onnx_helper.run(session, ai_path, None, \
                            {"gen_input_image": input_tiles}, return_first_output=True)

            for e in session_result:
                output_tiles.append(e)
```

Don't forget to save the file! After saving your changes, restart Siril to ensure the new script is loaded. You should now see your CoreML model appearing in the dropdown menu.

## Speed Test Results

Test Environment:

* Hardware: MacBook Air (M4 chip)
* Image Size: 3600 x 2700 pixels
* Condition: Real-world usage (several background applications open, but environment kept consistent for both tests).

Benchmark Results:

* ONNX Runtime (CPU Fallback): 314 seconds
* CoreML (GPU/ANE Accelerated): 46 seconds

Observations:

1. Performance Leap: The CoreML approach is approximately 6.8x faster, reducing processing time by nearly 85%.
2. Batch Size Efficiency: I tested this with Batch Size 1. While increasing the batch size might give the CPU better performance since it still has some margin, it’s not a big deal.
3. Apple Optimization: Interestingly, CoreML execution at Batch Size 1 appears to be faster than higher batch sizes, even when doubling the workload. It seems the hardware handles single-inference latency extremely well.

**Comment by m13v** [comment]
I hit this exact wall on an mlpackage conversion earlier this year. the RangeDim dynamic batch is probably why you see GPU not ANE, the neural engine needs static shapes, so dynamic-batch mlprograms silently fall back to GPU even with ComputeUnit.ALL. you can confirm it in Instruments with the Core ML template. enumeratedShapes([1, 4, 16, 32]) gets around it if you only need a few batch sizes. also the first inference after cold start eats ~500ms for compile cache warmup, worth pre-warming a dummy predict if latency matters at all.

**Comment by lonely-lockley** [comment]
@happpycorn Thank you for the report. I forked this repository and used Codex to apply the same approach to the main app itself. I added a model conversion script and patched the inference path.
Anyone can build the modified version by following the instructions in the README:
https://github.com/lonely-lockley/GraXpert

On my machine, the speedup is dramatic.

**Comment by cafuego** [comment]
To use the provided script to convert the model, make sure your venv uses the Mac-OS bundled python and not the latest one that homebrew installed. 

If you get `raise RuntimeError("BlobWriter not loaded")` when trying to run the script, this is your problem.
